cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java

```diff
@@ -48,56 +48,37 @@ private ConfigProducerUtil() {
+        String resolvedValue = resolveValue(name, config, defaultValue);
         if (hasCollection(injectionPoint.getType())) {
-            return convertValues(name, injectionPoint.getType(), getDefaultValue(injectionPoint), config);
+            return convertValues(name, injectionPoint.getType(), resolvedValue, defaultValue, config);
         }
 
-        return convertValue(name, resolveConverter(injectionPoint, config), getDefaultValue(injectionPoint), config);
+        return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(injectionPoint, config));
     }
 
     public static <T> T getValue(String name, Type type, String defaultValue, Config config) {
         if (name == null) {
             return null;
         }
-        return convertValue(name, resolveConverter(type, config), defaultValue, config);
+        String resolvedValue = resolveValue(name, config, defaultValue);
+        return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(type, config));
     }
 
-    public static <T> T convertValue(String name, Converter<T> converter, String defaultValue, Config config) {
-        String rawValue = getRawValue(name, config);
-        if (rawValue == null) {
-            rawValue = defaultValue;
-        }
-
-        T converted;
-        if (rawValue == null) {
-            // convert an empty value
-            try {
-                converted = converter.convert("");
-            } catch (IllegalArgumentException ignored) {
-                throw InjectionMessages.msg.propertyNotFound(name);
-            }
-        } else {
-            converted = converter.convert(rawValue);
-        }
-        if (converted == null) {
-            throw InjectionMessages.msg.propertyNotFound(name);
-        }
-        return converted;
-    }
-
-    public static <T> T convertValues(String name, Type type, String defaultValue, Config config) {
+    public static <T> T convertValues(String name, Type type, String resolvedValue, String defaultValue, Config config) {
         String rawValue = getRawValue(name, config);
         List<String> indexedProperties = ((SmallRyeConfig) config).getIndexedProperties(name);
         if (rawValue != null || indexedProperties.isEmpty()) {
-            return convertValue(name, resolveConverter(type, config), defaultValue, config);
+            return SmallRyeConfig.convertValue(name, resolvedValue, resolveConverter(type, config));
         }
 
         BiFunction<Converter<T>, IntFunction<Collection<T>>, Collection<T>> indexedConverter = (itemConverter,
                 collectionFactory) -> {
             Collection<T> collection = collectionFactory.apply(indexedProperties.size());
             for (String indexedProperty : indexedProperties) {
                 // Never null by the rules of converValue
-                collection.add(convertValue(indexedProperty, itemConverter, null, config));
+                collection.add(
```

Comment:
Absolutely. Great spot :)

Reply:
I think https://github.com/smallrye/smallrye-config/blob/fb4e21295a54a7053fcb5e0495da42275abc21da/cdi/src/main/java/io/smallrye/config/inject/ConfigProducerUtil.java#L81

Should probably also be 
```
SmallRyeConfig.convertValue(indexedProperty, resolveValue(indexedProperty, config, null), itemConverter));
```
since I don't think an individual `indexedProperty` can ever have a `defaultValue`

Reply:
Yes.